vowpalwabbit/csv_parser/src/parse_example_csv.cc

```diff
@@ -0,0 +1,280 @@
+    {
+      label_string = std::to_string(multiclass_label_counter.size() + 1);
+      multiclass_label_counter.insert(std::make_pair(label_content, multiclass_label_counter.size() + 1));
+    }
+    else
+    {
+      label_string = std::to_string(it->second);
+    }
+    label_content = VW::string_view(label_string);
+  }
+
+  all->example_parser->words.clear();
+  VW::tokenize(' ', label_content, all->example_parser->words);
+
+  if (!all->example_parser->words.empty())
+  {
+    all->example_parser->lbl_parser.parse_label(ae->l, ae->_reduction_features,
+        all->example_parser->parser_memory_to_reuse, all->sd->ldict.get(), all->example_parser->words, all->logger);
+  }
+}
+
+void parser::parse_namespaces(VW::workspace* all, example* ae, std::vector<VW::string_view> csv_line)
+{
+  // TODO: multiple namespaces support, currently using the filename as namespace
+  std::string ns = all->data_filename;
+  size_t found = ns.find_last_of('.');
+  if (found != std::string::npos) { ns = ns.substr(0, found); }
+
+  unsigned char _index = static_cast<unsigned char>(ns[0]);
+  bool new_index = false;
+
+  _channel_hash = all->example_parser->hasher(ns.data(), ns.length(), all->hash_seed);
+
+  if (ae->feature_space[_index].size() == 0) { new_index = true; }
+
+  ae->feature_space[_index].start_ns_extent(_channel_hash);
+  parse_features(all, ae->feature_space[_index], csv_line, (all->audit || all->hash_inv) ? ns.c_str() : nullptr);
+  if (new_index && ae->feature_space[_index].size() > 0) { ae->indices.push_back(_index); }
+
+  ae->feature_space[_index].end_ns_extent();
+}
+
+void parser::parse_features(VW::workspace* all, features& fs, std::vector<VW::string_view> csv_line, const char* ns)
+{
+  size_t _anon = 0;
+  for (size_t i = 0; i < csv_line.size(); i++)
+  {
+    // Skip label
+    if (label_index == i) { continue; }
+    // Ignore namespace info value
+    float _cur_channel_v = 1.f;
+    uint64_t word_hash;
+    float _v;
+    // Use header line for feature names
+    VW::string_view feature_name = _header[i];
+    VW::string_view string_feature_value = csv_line[i];
+    bool is_feature_float = check_if_float(string_feature_value);
+
+    float float_feature_value = 0.f;
+
+    if (is_feature_float)
+    {
+      float_feature_value = std::stof(string_feature_value.data());
+      _v = _cur_channel_v * float_feature_value;
+    }
+    else
+    {
+      _v = 1;
+    }
+
+    // Case where feature value is string
+    if (!is_feature_float)
+    {
+      // chain hash is hash(feature_value, hash(feature_name, namespace_hash)) & parse_mask
+      word_hash = (all->example_parser->hasher(string_feature_value.data(), string_feature_value.length(),
+                       all->example_parser->hasher(feature_name.data(), feature_name.length(), _channel_hash)) &
+          all->parse_mask);
+    }
+    // Case where feature value is float and feature name is not empty
+    else if (!feature_name.empty())
+    {
+      word_hash =
```

Comment:
It looks like you are relying on the fact that hashing a number results in getting that number back in the case of no header. This will break if a user supplies `--hash all`, instead of using the default.

https://vowpalwabbit.org/docs/vowpal_wabbit/python/latest/command_line_args.html#hash

It's also slower to go through the hash function when you know it can be avoided here.

Reply:
```cpp
    // Case where feature value is float and feature name is empty
    else
    {
      word_hash = _channel_hash + _anon++;
    }
```
Well, I guess you are actually talking about the lines above? (as of no header, which means the feature name is empty?) Do you mean that we should throw an error here for the CSV parser specifically? Could you please be more clear on how to avoid it? Thanks!

BTW, Actually I take the whole idea from here, I guess there should be no difference here for the situation you mentioned.
https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/vowpalwabbit/core/src/parse_example.cc#L197-L222

Reply:
Or maybe you mean that when there's no header, I use the number in order to replace the feature name? Well, actually those numbers are not really matters, and I just consider them as a string generator. It can be any string actually, I just use the number in order for the convenience and not happens to make two feature name the same.

And it works fine when I remove the header and run with `vw --csv --csv_no_header --hash all -a -d Iris.csv`

Reply:
![image](https://user-images.githubusercontent.com/43995067/168069685-b471cc8e-7e82-43e2-b02d-ab13abafafb9.png)
OK, Looks like I have some misunderstandings here. There's no need to distinguish those features with string names. So will just leave those features in empty.

> Or maybe you mean that when there's no header, I use the number in order to replace the feature name? Well, actually those numbers are not really matters, and I just consider them as a string generator to distinguish between different features. It can be any string actually, I just use the number in order for the convenience and not happens to make two feature name the same.
> 
> And it works fine when I remove the header and run with `vw --csv --csv_no_header --hash all -a -d Iris.csv` ![image](https://user-images.githubusercontent.com/43995067/168068247-9e372db9-8c3e-4467-b2b3-46bd31b78f3d.png)

Reply:
The hashed strings will result in an index to store the feature at. But VW has special logic for `--hash strings` for numbers. To demonstrate try this out:

1. Use acsv file no header and use your existing code which uses 1, 2, 3 etc 
2. Run `vw` with your data and add  `--invert_hash model.txt` to produce an invert_hash model. This will show you feature name, feature hash and model weight per feature
3. Repeat the same experiment using `--hash all` and `--hash strings` and look at the different feature hashes produced

So, in the case of `--hash strings` the numbers are just used an offset to the hash of the namespace name. This is the behavior we want if someone uses csv with no header. 

The reason I bring this up is I think we should not behave differently based on the value of the `--hash` option in the case of header-less CSV. By using a string `1` to `n` we are leaking an implementation detail I think we can hide and be more efficient at the same time.

I hope this helps clarify? Please let me know if it is still confusing.

Reply:
Yeah, that helps, thanks. Now I have already just leave all those feature names empty. https://github.com/HollowMan6/vowpal_wabbit/commit/df502be93b9637c0bdfbaffbe9021191a2a77bb7 It works OK because of this branch

https://github.com/HollowMan6/vowpal_wabbit/blob/df502be93b9637c0bdfbaffbe9021191a2a77bb7/vowpalwabbit/csv_parser/src/parse_example_csv.cc#L198-L202